File: matlab/src/api/snc/snc_env_tokenbucket.m
%{ @file snc_env_tokenbucket.m
 %  @brief Deterministic token-bucket arrival envelope
 %
 %  @author LINE Development Team
%}

%{
 % @brief Deterministic token-bucket arrival envelope
 %
 % @details
 % A flow policed by a (b,r) token bucket satisfies A(s,t) <= b + r*(t-s) with
 % probability one, hence E[exp(theta*A(s,t))] <= exp(theta*(r*(t-s)+b)) for
 % every theta > 0. The envelope is therefore constant in theta:
 %
 %   sigma(theta) = b,   rho(theta) = r.
 %
 % This is the deterministic network calculus arrival curve read as a degenerate
 % MGF envelope, so it can be mixed freely with the stochastic ones.
 %
 % @par Syntax:
 % @code
 % [sigma,rho] = snc_env_tokenbucket(b,r)
 % [sigma,rho] = snc_env_tokenbucket(b,r,theta)   % theta ignored, kept for
 %                                                % handle compatibility
 % @endcode
 %
 % @par Parameters:
 % <table>
 % <tr><th>Name<th>Description
 % <tr><td>b<td>Bucket depth, units of work
 % <tr><td>r<td>Token rate, work per slot
 % <tr><td>theta<td>Optional Chernoff parameter, unused
 % </table>
 %
 % @par Returns:
 % <table>
 % <tr><th>Name<th>Description
 % <tr><td>sigma<td>Burst term of the envelope, equal to b
 % <tr><td>rho<td>Rate term of the envelope, equal to r
 % </table>
 %
 % @par Reference:
 % J.-Y. Le Boudec and P. Thiran, "Network Calculus", LNCS 2050, Springer, 2001.
%}
function [sigma, rho] = snc_env_tokenbucket(b, r, theta)

if nargin < 2
    line_error(mfilename, 'Usage: [sigma,rho] = snc_env_tokenbucket(b,r).');
end
if b < 0 || r < 0
    line_error(mfilename, 'b and r must be nonnegative. Got %g, %g.', b, r);
end
if nargin >= 3 && theta <= 0
    line_error(mfilename, 'theta must be positive. Got %g.', theta);
end

sigma = b;
rho = r;
end

File: jar/src/main/java/jline/api/snc/Snc_env_tokenbucket.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */
package jline.api.snc;

/**
 * Deterministic token-bucket arrival envelope.
 *
 * <p>A flow policed by a (b,r) token bucket satisfies {@code A(s,t) <= b+r*(t-s)}
 * with probability one, so the envelope is constant in theta. This is the
 * deterministic network calculus arrival curve read as a degenerate MGF
 * envelope, so it mixes freely with the stochastic ones.</p>
 *
 * <p>Port of matlab/src/api/snc/snc_env_tokenbucket.m.</p>
 */
public final class Snc_env_tokenbucket {
    private Snc_env_tokenbucket() {}

    /**
     * @param b bucket depth, units of work
     * @param r token rate, work per slot
     * @return {b, r}
     */
    public static double[] snc_env_tokenbucket(double b, double r) {
        if (b < 0 || r < 0) {
            throw new IllegalArgumentException(
                    "snc_env_tokenbucket: b and r must be nonnegative, got " + b + ", " + r);
        }
        return new double[] {b, r};
    }

    /**
     * @param b     bucket depth
     * @param r     token rate
     * @param theta Chernoff parameter, accepted and ignored
     * @return {b, r}
     */
    public static double[] snc_env_tokenbucket(double b, double r, double theta) {
        if (theta <= 0) {
            throw new IllegalArgumentException("snc_env_tokenbucket: theta must be positive, got " + theta);
        }
        return snc_env_tokenbucket(b, r);
    }

    /**
     * @param b bucket depth
     * @param r token rate
     * @return the envelope as a function of theta
     */
    public static SncEnvelope of(final double b, final double r) {
        return new SncEnvelope() {
            public double[] eval(double theta) {
                return snc_env_tokenbucket(b, r, theta);
            }
        };
    }
}
